apache/metron
# METRON-2109: Add option to use Metron GUID as the id in Elasticsearch

## Contributor Comments
This PR adds the option described in the title.  A user can now configure a sensor (or all sensors) to use the Metron generated GUID instead of letting Elasticsearch generate one.  

### Changes Included
- Added a configuration option to the various indexing and writer configuration classes
- Added logic in the `ElasticsearchWriter` class to conditionally include the Metron GUID as the document id
- Added unit tests to cover this new use case
- Added documentation to the metron-indexing README explaining the setting

### Testing
This has been tested in full dev. 

1. Spin up full dev and verify data is showing up in Elasticsearch and the Alerts UI
2. Perform a basic search against the bro index:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{
  "from": 0,
  "indices": [
    "bro"
  ],
  "query": "*",
  "size": 5
}' 'http://user:password@node1:8082/api/v1/search/search'
```
The `id` fields should have an id that was generated by ES:
```
{
  "total": 93780,
  "results": [
    {
      "id": "AWqbP-_e08M1mJHcFnht",
      "source": {
    ...
```
3. Configure the `bro` sensor to use the Metron generated GUID instead:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{
  "hdfs": {
    "index": "bro",
    "batchSize": 1,
    "enabled": true
  },
  "elasticsearch": {
    "index": "bro",
    "batchSize": 1,
    "enabled": true,
    "metronId": true
  },
  "solr": {
    "index": "bro",
    "batchSize": 1,
    "enabled": true
  }
}' 'http://user:password@node1:8082/api/v1/sensor/indexing/config/bro'
```
4. Perform another search against `bro` and verify the `id` field is now the same as the Metron `guid` field:
```
{
  "total": 95475,
  "results": [
    {
      "id": "023ed7d1-3d1a-4530-bd1c-8a177e9a8323",
      "source": {
      ...
```
5. Perform a search against `snort` and verify the `id` field is still being generated by ES.
6. Set all sensors to use the Metron generated GUID in the global config:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{
    ...
    "indexing.writer.metronId": true
}' 'http://user:password@node1:8082/api/v1/global/config'
```
7. Verify searches against `bro`, `snort`, and `yaf` all return results with the `id` field set to the Metron `guid` field.
8. Navigate to the Alerts UI.  Verify messages (with and without the Metron generated GUID) can be updated by clicking on an alert and setting it's status to `OPEN` or any other status.

I also tested the REST endpoints directly and found no issues.  

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:


### For code changes:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```


### For documentation related changes:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.
